src/sidebar/store/modules/drafts.js

```diff
@@ -0,0 +1,194 @@
+    this.isPrivate = changes.isPrivate;
+    this.tags = changes.tags;
+    this.text = changes.text;
+  }
+  /**
+   * Returns true if `draft` is a draft for a given
+   * annotation.
+   *
+   * Annotations are matched by ID or local tag.
+   */
+  match(annotation) {
+    return (
+      (this.annotation.$tag && annotation.$tag === this.annotation.$tag) ||
+      (this.annotation.id && annotation.id === this.annotation.id)
+    );
+  }
+  /**
+   * Return true if this draft is empty and can be discarded without losing
+   * any user input.
+   */
+  isEmpty() {
+    return !this.text && this.tags.length === 0;
+  }
+}
+
+/* Reducer */
+
+const update = {
+  DISCARD_ALL_DRAFTS: function() {
+    return {
+      drafts: [],
+    };
+  },
+  REMOVE_DRAFT: function(state, action) {
+    const drafts = state.drafts.filter(draft => {
+      return !draft.match(action.annotation);
+    });
+    return {
+      drafts,
+    };
+  },
+  UPDATE_DRAFT: function(state, action) {
+    // removes a matching existing draft, then adds
+    const drafts = state.drafts.filter(draft => {
+      return !draft.match(action.draft.annotation);
+    });
+    drafts.push(action.draft); // push ok since its a copy
+    return {
+      drafts,
+    };
+  },
+  ADD_DRAFT: function(state, action) {
+    return {
+      drafts: state.drafts.concat(action.draft), // non-mutating
+    };
+  },
+};
+
+const actions = util.actionTypes(update);
+
+/* Actions */
+
+/**
+ * Create or update the draft version for a given annotation by
+ * replacing any existing draft or simply creating a new one.
+ */
+function createDraft(annotation, changes) {
+  return {
+    type: actions.UPDATE_DRAFT,
+    draft: new Draft(annotation, changes),
+  };
+}
+
+/** Remove all drafts. */
+function discardAllDrafts() {
+  return {
+    type: actions.DISCARD_ALL_DRAFTS,
+  };
+}
+
+/** Remove the draft version of an annotation. */
+function removeDraft(annotation) {
+  return {
+    type: actions.REMOVE_DRAFT,
+    annotation,
+  };
+}
+
+/* Selectors */
+
+/**
+ * Returns the number of drafts - both unsaved new annotations, and unsaved
+ * edits to saved annotations - currently stored.
+ *
+ * @return {number}
+ */
+function countDrafts(state) {
+  return state.drafts.length;
+}
+
+/**
+ * Retrieve the draft changes for an annotation.
+ *
+ * @return {Draft|null}
+ */
+function getDraft(state, annotation) {
+  for (let i = 0; i < state.drafts.length; i++) {
+    const draft = state.drafts[i];
+    if (draft.match(annotation)) {
+      return draft;
+    }
+  }
+
+  return null;
+}
+
+/**
+ * Returns the draft changes for an annotation, or null if no draft exists
+ * or the draft is empty.
+ *
+ * Since this is a non-primitive object injected into the store, there are
+ * some consequences with serialization. Slack discussion:
+ * https://hypothes-is.slack.com/archives/C1M8NH76X/p1562779754051700?thread_ts=1562776681.050500&cid=C1M8NH76X
```

Comment:
It is true we may need to do some extra work if we need to serialise state in future, but that's not really specific to this particular method, so I'm not sure this comment is useful here.